### ISSUE TITLE
Prioritize soft_oal in the router DLL / Make the router only load "wrap_oal" as a fallback

### ISSUE THREAD

**Issue by cammy8580** [kcat/openal-soft#976]
When using the router DLL, soft_oal will not be touched when a wrap_oal is present in the system folders. Games like Kane & Lynch 2: Dog Days will completely ignore soft_oal and instead load wrap_oal. 

Not many people know this, and some may report OAL Soft not working in titles that load a wrap_oal from the System32/SysWOW64 folder. 

If it's possible, it would be nice if the router DLL had a check to prioritize loading soft_oal, rather than loading wrap_oal to guarantee it always loading. It also could possibly be implemented into a toggleable option in the .ini as a bonus.

(Of course, you could probably just delete the wrap_oal from the folders, but still would be neat nonetheless.)

**Comment by kcat** [comment]
I'm not too keen on adding built-in prioritization of specific DLLs, since OpenAL Soft's router is intended to be vendor-neutral. Aside from prioritizing search paths (same path as the router dll, the current working directory, the executable/process path, then the system path), they're loaded in the order that `Find[First/Next]FileW` gets them. It aims to be thin and as hands-off as possible with the actual OpenAL drivers.

However, starting with commit 394e133b9fef1bfbde3ee5842b489bf8e91a135b, you can use the `ALROUTER_ACCEPT` and/or `ALROUTER_REJECT` environment variables to specify (comma-separated) dll names it can/can't load. For instance, setting `ALROUTER_ACCEPT=soft_oal.dll` will allow the router to load only `soft_oal.dll` as found in any of the searched paths, while `ALROUTER_REJECT=wrap_oal.dll,ct_oal.dll` will allow the router to load any dlls that aren't `wrap_oal.dll` or `ct_oal.dll` in any of the searched paths.

I do intend to do more for user-configuration of the router's behavior, including custom search paths and user-specified driver prioritization, but it's not there yet.

**Comment by falafelrapper** [comment]
Out of curiousity, was there ever a fix to find out how to get OpenAL soft working with Kane and Lynch 2 in particular? I myself have gotten OpenAL soft working with every game, and tried multiple different ways such as removing wrap_oal, replacing wrap_oal with soft_oal and renaming it, replacing OpenAL32 with soft_oal being renamed to it, and none of it seems to work. Would love maybe some help on what some recommendations people have, as I would love to use my surround sound system with this game.

**Comment by mirh** [comment]
https://www.youtube.com/watch?v=vENnMnUFZxA
Considering the game [relies](https://steamdb.info/depot/28001/) on the official openal redist, it seems [pretty](https://github.com/kcat/openal-soft/issues/650#issuecomment-1186326473) [normal](https://github.com/kcat/openal-soft/issues/757#issuecomment-1231616025) for openal-soft not to be picked up.
But replacing openal32.dll still not being enough is very very strange.

**Comment by mirh** [comment]
I checked the exe in ghidra and I couldn't really put all pieces together... The function that eventually calls *alcOpenDevice* was left "dangling" and unreferenced from the rest of the program. 

But I can tell that that around the place where they parse some of the settings (`NoEnvMorph`, `EnableOpenAL`, `EnableDieselPower`, `NumberOfStreams`) there's an explicit "Generic Software" mention.
Soo.. Like, I very much couldn't tell if that's used directly or if there couldn't be some other logic involved (so maybe it wouldn't hurt to check with a X-Fi, or with the good old "exact device name matching" trick) but there are chances that they are just explicitly asking for that.

**Comment by falafelrapper** [comment]
Sorry to bring this back up, but I've been looking around and maybe I've overlooked something for the recommendations above. 

Do you know where I could find instructions on the "exact device name matching" trick? I don't have an x-fi sound card, but would love to give this trick a test in my free time, K&L2 is a favorite and have been wanting to finally get the surround sound track working if possible, so if this is the fix, that would be awesome!

**Comment by mirh** [comment]
https://github.com/kcat/openal-soft/issues/1001#issuecomment-2328619188
https://github.com/kcat/dsoal/issues/51#issuecomment-2378498509
Long story short just about every openal selection problem can be solved by having its alcDefaultName match the device name as reported by Windows.

The thing is that sometimes games can do very stupid things on their own because the developers were clueless AF (in this case it's not clear if they added a specific mention to "generic software" just as an additional explicit fallback, or to forcefully use it always).
Though if @IDrinkLava could get (at least the demo AFAIU?) to work without much mindblowing trickery I guess it can't be that bad.


p.s. just realized OP is the guy I was criticizing [here](https://github.com/kcat/openal-soft/pull/1008#issuecomment-2564760979)

**Comment by falafelrapper** [comment]
Would this be through manually building my own file or by just doing alcDefaultName in the alsoft.ini? Sorry for all the questions, I've been going through a rabbit hole the past couple of months trying to figure this out and feel like I keep looking in the wrong places lol

**Comment by mirh** [comment]
As mentioned in the video as far as K&L2 goes, just replacing openal32.dll with openal-soft (essentially making it the only driver available) should be already enough.
The "proper" way (assuming the game isn't screwed) as mentioned in my links is instead to recompile the dll after changing [this](https://github.com/kcat/openal-soft/blob/1.24.3/alc/alc.cpp#L289).

**Comment by Firedavinci** [comment]
Any solutions yet for Kane and Lynch 2 and OpenAL Soft? @kcat 

I just downloaded the GOG version of the game, but the game only works with the original OpenAL dll's, specifically OpenAL32.dll and wrap_oal.dll. Both have to be in the folder for the game to start. I was able to make it work with just the wrap_oal.dll renamed to OpenAl32.dll. However the original OpenAl32.dll by itself won't work, it needs the wrap_oal.dll for the game to start.

Regarding OpenAl Soft, I've tried all possible variations with soft_oal.dll trying to get the game running, but it just won't start. 

Here is the log for Kane and Lynch2:

[ALSOFT] (II) Initializing library v1.25.1-c41d64c6 master
[ALSOFT] (II) Supported backends: wasapi, dsound, winmm, jack, null, wave
[ALSOFT] (II) Got binary: F:\Kane & Lynch 2 Dog Days, kl2.exe
[ALSOFT] (II) Loading config F:\Kane & Lynch 2 Dog Days\alsoft.ini...
[ALSOFT] (II) Vendor ID: ""
[ALSOFT] (II) Starting watcher thread
[ALSOFT] (II) Initialized backend "wasapi"
[ALSOFT] (II) Added "wasapi" for playback
[ALSOFT] (II) Added "wasapi" for capture
[ALSOFT] (II) Opening **playback device "Generic Software"**
[ALSOFT] (II) Got device "Speakers (ASUS Xonar Essence STX Audio Device)", "{177F5405-294A-4742-B49F-3138D2A0A75D}", "{0.0.0.00000000}.{177f5405-294a-4742-b49f-3138d2a0a75d}"
[ALSOFT] (II) Watcher thread started
[ALSOFT] (WW) Failed to find device name matching "Generic Software"
[ALSOFT] (WW) Failed to open playback device: Device "Generic Software" not found
[ALSOFT] (WW) Error generated on device 0x0, code 0xa004
[ALSOFT] (II) Freeing device 0x2cf0370
[ALSOFT] (WW) Querying error state on null context (implicitly 0xa004)


In contrast, here is the log for Quake 4 which is using soft_oal renamed to OpenAl32.dll:

[ALSOFT] (II) Initializing library v1.25.1-c41d64c6 master
[ALSOFT] (II) Supported backends: wasapi, dsound, winmm, jack, null, wave
[ALSOFT] (II) Got binary: F:\Quake 4, Quake4 - Run.exe
[ALSOFT] (II) Loading config F:\Quake 4\alsoft.ini...
[ALSOFT] (II) Vendor ID: ""
[ALSOFT] (II) Starting watcher thread
[ALSOFT] (II) Initialized backend "wasapi"
[ALSOFT] (II) Added "wasapi" for playback
[ALSOFT] (II) Added "wasapi" for capture
[ALSOFT] (II) Opening **default playback device**
[ALSOFT] (II) Got device "Speakers (ASUS Xonar Essence STX Audio Device)", "{177F5405-294A-4742-B49F-3138D2A0A75D}", "{0.0.0.00000000}.{177f5405-294a-4742-b49f-3138d2a0a75d}"
[ALSOFT] (II) Watcher thread started
[ALSOFT] (II) Created device 0x96c4c00, "OpenAL Soft on Speakers (ASUS Xonar Essence STX Audio Device)"

**Comment by Firedavinci** [comment]
> I checked the exe in ghidra and I couldn't really put all pieces together... The function that eventually calls _alcOpenDevice_ was left "dangling" and unreferenced from the rest of the program.
> 
> But I can tell that that around the place where they parse some of the settings (`NoEnvMorph`, `EnableOpenAL`, `EnableDieselPower`, `NumberOfStreams`) there's an explicit "Generic Software" mention. Soo.. Like, I very much couldn't tell if that's used directly or if there couldn't be some other logic involved (so maybe it wouldn't hurt to check with a X-Fi, or with the good old "exact device name matching" trick) but there are chances that they are just explicitly asking for that.

You are correct @mirh , the openal soft log also mentions it: 
[ALSOFT] (WW) Failed to find device name matching "Generic Software"
[ALSOFT] (WW) Failed to open playback device: Device "Generic Software" not found

What can be done in this case, if any? You said something about  "exact device name matching trick, would that work and how can I do it?

**Comment by kcat** [comment]
If there's no way to change the device name that the game tries to open, the only solution would be to modify OpenAL Soft or the router to explicitly check for `"Generic Software"` in `alcOpenDevice` and ignore it (open the default instead). And hope the game doesn't check which device opened and complain/error if it doesn't report back `"Generic Software"`.

Maybe you could alternatively hex-edit the executable (make sure to back it up first, so you can restore it) and change mentions of `Generic Software` to `OpenAL Soft` (adding extra nul chars to fill the extra characters). Though that does risk breaking if other parts of the executable reference parts of that string where it needs the same characters. e.g. if one part of the executable uses the string `"Generic Software"`, and another uses the string `"Software"`, the compiler can combine them and change references to the latter string to the former string with an 8-byte offset; making that change would then make any references to the string `"Software"` become `"oft"`, if that happened.

**Comment by Firedavinci** [comment]
> Maybe you could alternatively hex-edit the executable (make sure to back it up first, so you can restore it) and change mentions of `Generic Software` to `OpenAL Soft` (adding extra nul chars to fill the extra characters). Though that does risk breaking if other parts of the executable reference parts of that string where it needs the same characters. e.g. if one part of the executable uses the string `"Generic Software"`, and another uses the string `"Software"`, the compiler can combine them and change references to the latter string to the former string with an 8-byte offset; making that change would then make any references to the string `"Software"` become `"oft"`, if that happened.

It's working!!! Replacing "Generic Software" in Hex editor with "OpenAL Soft" did the trick. Thank you man. Surround sound is working now too. So satisfying to get this game to run with OpenAL Soft finally. @cammy8580 @falafelrapper

**Comment by cammysorammy** [comment]
> It's working!!! Replacing "Generic Software" in Hex editor with "OpenAL Soft" did the trick. Thank you man. Surround sound is working now too. So satisfying to get this game to run with OpenAL Soft finally. [@cammy8580](https://github.com/cammy8580) [@falafelrapper](https://github.com/falafelrapper)

Sick, nice to see it's working out! Gonna see if I can try to get this integrated into KL2Fix.

**Comment by cammysorammy** [comment]
It looks like unfortunately there are some crashes when you do this method, but I'll keep you guys posted if I do find anything to circumvent this.

**Comment by Firedavinci** [comment]
> It looks like unfortunately there are some crashes when you do this method, but I'll keep you guys posted if I do find anything to circumvent this.

You mean crashes at start or mid game? I had crashes at first (game didnt start) because I didnt correctly replace the word "Generic Software" with the correct amount of nul and space. Now its working for me. Haven't tested it extensively though, only 5 min.

Edit
Hex:
4f 70 65 6e 41 4c 20 53 6f 66 74 00 00 00 00

**Comment by cammysorammy** [comment]
> > It looks like unfortunately there are some crashes when you do this method, but I'll keep you guys posted if I do find anything to circumvent this.
> 
> You mean crashes at start or mid game? I had crashes at first (game didnt start) because I didnt correctly replace the word "Generic Software" with the correct amount of nul and space. Now its working for me. Haven't tested it extensively though, only 5 min.
> 
> Edit Hex: 4f 70 65 6e 41 4c 20 53 6f 66 74 00 00 00

Might have goofed something up, I'll take a look again and let you know.

**Comment by Firedavinci** [comment]
> Might have goofed something up, I'll take a look again and let you know.

I tested it for longer now and all seems fine, no crashes. However couple of things:

1. the game seems to have reversed y audio (up & down) adding "reverse-y = true" in alsoft.ini fixes that. 
2. efx/eax is very loud, I would suggest "boost=-12".

**Comment by cammysorammy** [comment]
Yeah I have no clue why, but it seems like I'm having a crash every time I attempt to headshot an enemy. I'll attach a log here, but we should probably take this to another issue.

[alsoft_error.txt](https://github.com/user-attachments/files/25221503/alsoft_error.txt)

**Comment by Firedavinci** [comment]
> Yeah I have no clue why, but it seems like I'm having a crash every time I attempt to headshot an enemy. I'll attach a log here, but we should probably take this to another issue.
> 
> [alsoft_error.txt](https://github.com/user-attachments/files/25221503/alsoft_error.txt)

Do you have any other mods running? I have the GOG version,  maybe the Steam .exe is different? 
Lastly, does the game start when using soft_oal.dll as OpenAl32.dll without the HEX mod? 
Why create a new issue, this is the same topic no?

**Comment by mirh** [comment]
The latest ghidra seems to be a lil better with decompiling, even though I couldn't make a sense of all those nameless variables. 
There's supposedly options called "SoundRender" and "SoundDevice"~~, which may or may not do anything in the code.
I'm not sure where they could be queried from? Judging from the other ones surrounding them, they may [come](https://github.com/HHCHunter/Hitman-BloodMoney/blob/master/CONFIG-FILES/HitmanBloodMoney.ini#L116) from an ini? The game seems to try to open `KL2.ini` (couldn't find referenced anywhere in the code) and `main.ini` in its folder but they aren't there.~~

Did you try to hex edit "Generic Software" inside the `0.kls` save game?
Steam version is CEG protected and only the gods know the quirks that may cause with modifications.

**Comment by cammysorammy** [comment]
> > Yeah I have no clue why, but it seems like I'm having a crash every time I attempt to headshot an enemy. I'll attach a log here, but we should probably take this to another issue.
> > [alsoft_error.txt](https://github.com/user-attachments/files/25221503/alsoft_error.txt)
> 
> Do you have any other mods running? I have the GOG version, maybe the Steam .exe is different? Lastly, does the game start when using soft_oal.dll as OpenAl32.dll without the HEX mod?

I've tested it on both the GOG and steam EXE, plus with KL2Fix installed and without it. I really have no clue as to what's happening here.

**Comment by Firedavinci** [comment]
> I've tested it on both the GOG and steam EXE, plus with KL2Fix installed and without it. I really have no clue as to what's happening here.

Try my GOG .exe it is already modified for OpenAL Soft. If this doesn't work, I would suggest deleting the save folder and registry for KL2, then try again. 
Edit. not sure but try enabling Vsync if you havent.

[kl2.zip](https://github.com/user-attachments/files/25224881/kl2.zip)

**Comment by cammysorammy** [comment]
> > I've tested it on both the GOG and steam EXE, plus with KL2Fix installed and without it. I really have no clue as to what's happening here.
> 
> Try my GOG .exe it is already modified for OpenAL Soft. If this doesn't work, I would suggest deleting the save folder and registry for KL2, then try again. Edit. not sure but try enabling Vsync if you havent.
> 
> [kl2.zip](https://github.com/user-attachments/files/25224881/kl2.zip)

I had talked with 3DJ, and it seems like his specific compiled build of OAL soft that changes it so it uses Generic Software actually doesn't crash now, and he's made an issue for a feature request to make it so the name of the router/dlls can be altered in an .ini. see this issue here: #1243 

Much better solution than altering anything in the executable imho

**Comment by cammysorammy** [comment]
> 2. efx/eax is very loud, I would suggest "boost=-12".

btw, i think this is intentional. multiple levels actually don't reverberate properly on the PC version, which OAL soft fixes with the proper configuration of it.

**Comment by mirh** [comment]
You know I mentioned above you can try editing it in the saves, right?

**Comment by Firedavinci** [comment]
> actually doesn't crash now, and he's made an issue for a feature request to make it so the name of the router/dlls can be altered in an .ini. see this issue here: [#1243](https://github.com/kcat/openal-soft/issues/1243)
> 
> Much better solution than altering anything in the executable imho

Great! And good idea to have that configurable in the INI instead of the EXE, however it was kinda fun to solve this the hard way lol.

**Comment by cammysorammy** [comment]
> You know I mentioned above you can try editing it in the saves, right?

Yeah, and when I had changed it in the .exe it had changed it there too. It made no difference. Regardless, it's still much easier and potentially could assist with supporting more games now that a proper feature request for changing these in a .ini is present.

**Comment by mirh** [comment]
@Tervel1337 would you be able to comment on the logic for setting what I believe is the hidden _SoundDevice_ variable?
Is it hardcoded (and then at most just nominally written in the kls file, but never read) or is there more to the eye than it seems? Is there any reason changing that one exe string (or any?) would cause crashes? Are the game's .ini files workable for an override?

**Comment by cammysorammy** [comment]
Tervel had made a specific build that had made the exact same changes as the hex-edit when we were testing it and found that (at least on my rig), that when the game has that string, for some reason, at least on my rig the game crashes with this log. At no point does the line in the .kls save need to be altered at all. The executable already does it FOR you.

[KL2.exe.20260211161908.log](https://github.com/user-attachments/files/25247740/KL2.exe.20260211161908.log)

I just looked into it however and it looks like that there was a bug in his fix with the `UncapLoadScreen` toggle being set to 1 (experimental option to load levels with an uncapped framerate), which I had helped him debug earlier, and now after testing the same fix with the steam version, it looks like the crash is no longer occurring. This has already been fixed in his repo.

Regardless, I do still think that OpenAL Soft should have that feature mentioned in that issue for the sake of convenience, as if something similar can be implemented like the `version-override`, `vendor-override`, and `renderer-override` for the purpose of changing which device the game attempts to grab, it would be a lot more convenient to just change one line in an .ini than wait for someone to make a hex edit or a .asi patch for games that do not natively launch with any other device name.

**Comment by mirh** [comment]
Yes, I know (and also thanks god it wasn't the string to create problems).
It's just that regardless of that, for the moment I'd like to treat this game as an outlier (because every other single time there was more to the story.. and you could argue that if this is really how it looks, then the game wasn't even ever meant to work differently).
And a dedicated fix existing for it seems just the perfect thing to wish away this uneasy dissonance.

**Comment by cammysorammy** [comment]
I'm fairly confident that the game behaves this way because with Kane & Lynch: Dead Men, they had actually given you the option to select your audio device to be anything you wished (it also supported the ability to use DirectSound, which you can still find references to in the K&L2 .exe):

<img height="300" alt="Image" src="https://github.com/user-attachments/assets/965f2daa-f823-41f7-b522-b7014e93a9cc" />

<img  height="300" alt="Image" src="https://github.com/user-attachments/assets/1254f130-12ce-4c40-87bb-b7154a6826d3" />

<img height="300" alt="Image" src="https://github.com/user-attachments/assets/efec06b3-02cf-4335-b4b5-bdc9b98a6951" />

Because of this, I wouldn't argue that this was how the game was meant to be done, and instead I attribute this to the port for K&L2 being outsourced to Nixxes Software rather than being handled in a similar fashion to the K&L1 port, which was handled by IOI themselves. Considering everything else that I have found out when attempting to assist in development for KL2Fix, it only leads me to believe that this happened because of a haphazard oversight done by Nixxes when they had done the port job, as with what was common at the time in the 7th console generation, and clearly evident all throughout this port.

I've discussed this with Tervel about incorporating something to change this on-the-fly in his .asi, and potentially include OAL soft inside of it as well, due to the benefits him and I had discovered of OAL Soft actually fixing reverb that was missing on the PC port. Still, I don't think that including an option to override in OAL Soft would be a bad idea, especially that as mentioned in #1243, there are other benefits to being able to incorporate something like this.

**Comment by Firedavinci** [comment]
> I'm fairly confident that the game behaves this way because with Kane & Lynch: Dead Men, they had actually given you the option to select your audio device to be anything you wished (it also supported the ability to use DirectSound, which you can still find references to in the K&L2 .exe):

I wonder does the game support only software audio? Because for one it expects the "Generic Software" and as I can see from your screenshots of KL1, it says Direct Sound software / Generic Software. Or does it not matter when replacing it with Openal Soft? I remember one of those games had to be toggled to hardware instead of software for the effects and surround sound to kick in.

**Comment by cammysorammy** [comment]
> I wonder does the game support only software audio? Because for one it expects the "Generic Software" and as I can see from your screenshots of KL1, it says Direct Sound software / Generic Software. Or does it not matter when replacing it with Openal Soft? I remember one of those games had to be toggled to hardware instead of software for the effects and surround sound to kick in.

I'm not exactly sure. The game supports both OpenAL and DirectSound to my knowledge, so it's just a matter of whatever implementation is better most likely. Might be worth to check using OALS and DSOAL and seeing which one is handled better.

**Comment by mirh** [comment]
> it also supported the ability to use DirectSound, which you can still find references to in the K&L2.exe

Idk how much it's "just references", and how much it's instead Nixxes being idiots (I agree with your analysis).
There's literally a `EnableDirectSound` between all the strings that I mentioned above, and something was def compiled in it if I could also find the "ZSynthDS::Initialize - Create DS-Object failed" error being present in a check after loading dsound.dll.

But as I said go figure the decompilation.. in an undefined function (the one that should also be supposed to finally call `alcOpenDevice`) there's a `alGetEnumValue("AL_FORMAT_QUAD16")` variable. Does it mean the game surround is actually 4.0?

> I've discussed this with Tervel about incorporating something to change this on-the-fly in his .asi, and potentially include OAL soft inside of it as well

Of course openal-soft is better than the default (and after much thinking, I realized openal is still probably the preferred api as that's probably what they used on xbox.. even though go figure about the ps3).

**Comment by falafelrapper** [comment]
I'm about 2 weeks late, this issue has been driving me nuts for the past year (no joke would some days go down a rabbit hole every couple of months only to spend like 5-6 hours losing my mind on why it wouldnt work and almost even broke all OpenAL soft games on my PC LMAO), only to see that both there's a KL2Fix community patch AND someone made an OpenAL Soft build that fixes this? Legitimately so hyped right now, I plugged in the OpenAL soft router from 3DJ and it immediately was working on my atmos HT setup when running windows in 7.1 surround. Hopefully though we can still get that override feature in the future in case there are other games hardcoded like this as well, instead of needing to look through forums just to find one specific build. Props though to @cammy8580 @ThreeDeeJay for their work on this and also the KL2Fix with @Terval1337

**Comment by ThreeDeeJay** [comment]
@falafelrapper Now that you mention it, does my custom soft_oal.dll get loaded when wrap_oal.dll is also enumerated?
I only tested replacing the router (OpenAL32.dll) directly which completely bypasses all other OpenAL libraries so I'm not sure if the game would still prefer OpenAL Soft's "Generic Software" over wrap_oal's in KL2 (or any other game, for that matter). 🤔  

On a side note, have you considered using OpenAL Soft's 7.1.4(.4?) mode to also get verticality if you have height speakers?
https://github.com/kcat/openal-soft/commit/e9ad8571ba93dd6631a9c05a05a28ede95728d9e
https://github.com/kcat/openal-soft/commit/6fa35608bcb21ebedcf683949c6fb20e4a9e3b0b

**Comment by mirh** [comment]
It would prefer it in the game folder, and it would loose if it's in the system folder (not 100% sure if you just deleted openal32.dll and it loaded the system one, but eh kinda accademically navel gazy 🤷)
